tests/unit/test_commands.py

```diff
@@ -822,56 +829,72 @@ def test_parse_args_upgrade(args, expected_CLIargs):
     assert parsed_args == expected_CLIargs
 
 
-@pytest.mark.parametrize(
-    "args",
-    [
-        ["upgrade", "hypervisors", "--machine 1", "--az 2"],
-        ["upgrade", "hypervisors", "--az 1", "-m 2"],
-    ],
-)
-def test_parse_args_hypervisors_exclusive_options(args):
+@patch("cou.commands.argparse.ArgumentParser.error")
+def test_parse_args_hypervisors_exclusive_options(mock_error):
     """Test parsing mutually exclusive hypervisors specific options."""
-    with pytest.raises(SystemExit, match="2"):
-        commands.parse_args(args)
+    mock_error.side_effect = SystemExit
+    with pytest.raises(SystemExit):
+        commands.parse_args(["upgrade", "hypervisors", "--machine", "1", "--az", "2"])
+    mock_error.assert_called_once_with(
+        "argument --availability-zone/--az: not allowed with argument --machine/-m"
+    )
+
+
+@patch("cou.commands.argparse.ArgumentParser.error")
+def test_parse_args_hypervisors_exclusive_options_reverse_order(mock_error):
+    """Test parsing mutually exclusive hypervisors specific options (reverse order)."""
+    mock_error.side_effect = SystemExit
+    with pytest.raises(SystemExit):
+        commands.parse_args(["upgrade", "hypervisors", "--az", "1", "-m", "2"])
+    mock_error.assert_called_once_with(
+        "argument --machine/-m: not allowed with argument --availability-zone/--az"
+    )
```

Comment:
Why we are mocking the `error` here and then catch what we set? The [error](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.error) function will raise SystemExit with return code 2 and we should check that instead of our mocked version.

Reply:
I had two main goals with this specific design:

1. have a more useful assert (previously it was asserting that SystemExit was raised, which could be caused by anything - case in point, some tests had typos, so the SystemExit was raised because of the function crashing on the typos, not because of the invalid condition that was supposed to be tested)
2. retain the original control flow of the function (if the mock doesn't raise SystemExit, then the function will continue instead of exiting). See https://github.com/canonical/charmed-openstack-upgrader/pull/580#discussion_r1791938631 for background.

Reply:
> which could be caused by anything

We expect that the error call sys.exit(2), so we tested it. If there is any type the error will not be SystemExit, but something else.

> retain the original control flow of the function (if the mock doesn't raise SystemExit, then the function will continue instead of exiting).

You are mocking `error` function, which if we call it will raise an exception. So you do not need to mock it at all.

Reply:
> We expect that the error call sys.exit(2), so we tested it. If there is any type the error will not be SystemExit, but something else.

So for example, `test_parse_args_quiet_verbose_exclusive()` was an invalid test. It was designed to verify the quiet and verbose flags are mutually exclusive, however the args used to test were invalid for other reasons (missing the plan or upgrade subcommand). The test passed because the function raised SystemExit as expected, but that same test would still pass if we removed the mutually exclusive rules from quiet/verbose.

If we're going to test this case, we need a better assert.